# 4주차 랩 · 회원과 비회원은 비·눈에 똑같이 줄어들까

**⏱ 60분** · 교재 4-1·4-2와 같은 Capital Bikeshare 2011~2012년 시간별 기록 · 개인 제출

## 🎯 이번 랩의 질문

교재에서는 비·눈이 오는 시간의 전체 대여를 평시의 **0.53배**로 계획하기로 했습니다. 그런데 전체 대여 `cnt`는 회원 `registered`와 비회원 `casual`의 합입니다. 두 집단이 비·눈에 다르게 반응한다면, 비회원이 많은 시간과 회원이 많은 시간의 계획을 같은 배수로 잡기 어렵습니다.

> **회원과 비회원 중 비·눈에 더 크게 줄어드는 쪽은 어디일까요? 두 집단의 계획 배수를 따로 둬야 할까요?**

교재에서 반복한 두 패턴을 회원과 비회원에 한 번씩 다시 씁니다.

| 패턴 | 교재 | 이번 랩 |
|---|---|---|
| 묶어서 평균 → 비·눈 ÷ 평시 | 4-1 | 과제 1·2 |
| 회귀 두 줄 — `smf.ols(...).fit()` → `np.exp(모형.params["rain"])` | 4-2 | 과제 3~6 |
| 숫자를 근거로 판단 쓰기 | 4-2 7절 | 과제 7 |

### 진행 방법

| 표시 | 뜻 | 할 일 |
|---|---|---|
| ▶ | 제공 코드 | 그대로 실행합니다 |
| 📝 | 과제 | `None` 자리를 코드로 바꿔 채웁니다. 변수 이름은 바꾸지 않습니다 |
| ✅ | 확인 셀 | 고치지 말고 실행만 합니다 |
| 💬 | 결과 읽기 | 내 결과에서 무엇을 볼지 짚습니다. 과제 7에서 이 숫자들을 씁니다 |

- 준비 셀부터 순서대로 `Shift + Enter`로 실행합니다. 결과는 반올림하지 않고 변수에 저장합니다.
- **✅ 확인 셀의 "형식 확인 완료"는 정답이라는 뜻이 아닙니다.** 작성 여부와 자료형만 봅니다. 계산값은 제출 후 교수자가 채점하고, 과제 7의 서술은 따로 읽습니다.
- 시간: 준비 5분 · 과제 1 6분 · 과제 2 5분 · 과제 3 7분 · 과제 4 5분 · 구간 확인 2분 · 과제 5 8분 · 과제 6 6분 · 과제 7 15분 · 제출 전 확인 1분

In [3]:
# ▶ 준비 셀 — 그대로 실행하세요. 교재와 같은 기록을 읽고 temp_c, rain 열을 만듭니다.
import os
import sys
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

for folder in (".", "..", "../..", "../../.."):
    if os.path.exists(os.path.join(folder, "balab.py")):
        sys.path.insert(0, folder)
        break
else:
    import urllib.request
    urllib.request.urlretrieve(
        "https://raw.githubusercontent.com/BALAB-PKNU/bizanalytics/main/balab.py", "balab.py")

from balab import load, checker, summary, check_format
check = checker("week04")
h = load("bike")
h["temp_c"] = h["temp"] * 47 - 8
h["rain"] = (h["weathersit"] >= 3).astype(int)
h[["dteday", "hr", "rain", "registered", "casual"]].head()

,dteday,hr,rain,registered,casual
0,2011-01-01,0,0,13,3
1,2011-01-01,1,0,32,8
2,2011-01-01,2,0,27,5
3,2011-01-01,3,0,10,3
4,2011-01-01,4,0,1,0


✅ 표의 앞 다섯 행이 나오면 됩니다. `registered`와 `casual`을 더하면 교재의 `cnt`입니다.

---
## 📝 과제 1 · 평균끼리 나누면 회원과 비회원은 각각 몇 배인가

가장 쉬운 답부터 구합니다. 교재 4-1 4절에서 전체 대여로 한 **묶어서 평균 → 비·눈 ÷ 평시**를 회원과 비회원에 따로 합니다. 뒤에서 조건을 맞춘 배수와 비교할 출발점입니다.

- `rain`으로 묶어 `registered`와 `casual` 두 열의 평균을 구하고 `means`에 담습니다. 두 열은 교재 4-1 6절처럼 리스트로 함께 고릅니다.
- 교재 4-1 4절처럼 `loc`으로 비·눈(행 1)의 평균을 평시(행 0)의 평균으로 나눕니다. 회원의 배수는 `reg_mean_ratio`, 비회원의 배수는 `cas_mean_ratio`에 담습니다.

**결과 형식:** 실수 두 개

In [7]:
# 📝 과제 1 — None 자리를 채우세요
# 힌트: 교재 4-1 4절의 두 셀과 같은 순서입니다. 묶은 뒤 cnt 대신 두 열을 고르고, 나눌 때는 열 이름을 registered 또는 casual로 씁니다.
means = h.groupby("rain")[["registered","casual"]].mean()           # rain별 registered·casual 평균 표
reg_mean_ratio = means.loc[1, "registered"]/means.loc[0,"registered"]  # 회원: 비·눈 평균 ÷ 평시 평균
cas_mean_ratio = means.loc[1, "casual"]/means.loc[0,"casual"]   # 비회원: 비·눈 평균 ÷ 평시 평균

(np.float64(0.6005233484969172), np.float64(0.42822924527817996))

In [5]:
# ✅ 확인 — 과제 1 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(1, reg_mean_ratio=(reg_mean_ratio, 0.001), cas_mean_ratio=(cas_mean_ratio, 0.001))
else:
    check_format(1, reg_mean_ratio=reg_mean_ratio, cas_mean_ratio=cas_mean_ratio)

형식 확인 완료: 과제 1 — 계산값의 정답 여부는 제출 후 채점합니다.


💬 **결과 읽기** — 두 배수 중 어느 쪽이 더 작나요? 교재의 전체 대여 평균끼리 배수 0.57과도 비교해 보세요.

---
## 📝 과제 2 · 비회원은 대여 0건인 시간이 얼마나 되나

과제 3부터는 교재 4-2처럼 대여 건수에 로그를 씌워 회귀합니다. 그런데 0에는 로그를 씌울 수 없습니다. 준비 셀과 같은 기록에서 회원이 0건인 시간은 24시간뿐이지만 비회원은 1,581시간입니다. 0건 시간이 날씨에 따라 다르게 나타나는지 먼저 봅니다.

- `casual`이 0이면 1, 아니면 0인 열 `cas_zero`를 만듭니다. 교재 4-1 2절에서 `rain`을 만든 방법(비교 → `astype(int)`)과 같습니다.
- `rain`으로 묶어 `cas_zero`의 평균을 구합니다. 교재 4-1 5절처럼 0·1 열의 평균은 1의 비율, 곧 0건 시간의 비율입니다. 평시(rain 0)의 비율은 `cas_zero_dry`, 비·눈(rain 1)의 비율은 `cas_zero_rain`에 담습니다.

**결과 형식:** 실수 두 개

In [8]:
# 📝 과제 2 — None 자리를 채우세요
# 힌트: 교재 4-1 2절의 rain 만들기에서 비교를 "casual이 0과 같은가"(==)로 바꿉니다. 비율은 교재 4-1 5절처럼 묶어서 평균입니다.
h["cas_zero"] = (h["casual"] == 0).astype(int)  # casual이 0이면 1, 아니면 0
cas_zero_dry = h.groupby("rain")["cas_zero"].mean().loc[0]   # 평시(rain 0) 중 비회원 0건 시간의 비율
cas_zero_rain = h.groupby("rain")["cas_zero"].mean().loc[1]  # 비·눈(rain 1) 중 비회원 0건 시간의 비율

In [9]:
# ✅ 확인 — 과제 2 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(2, cas_zero_dry=(cas_zero_dry, 0.001), cas_zero_rain=(cas_zero_rain, 0.001))
else:
    check_format(2, cas_zero_dry=cas_zero_dry, cas_zero_rain=cas_zero_rain)

형식 확인 완료: 과제 2 — 계산값의 정답 여부는 제출 후 채점합니다.


💬 **결과 읽기** — 비·눈 시간의 0건 비율이 평시와 다르다면, 0건 시간을 빼고 로그를 씌울 때 평시와 비·눈 시간이 서로 다른 비율로 빠집니다. 그래서 과제 3·4에서는 시간을 빼지 않고 대여에 1을 더해 로그를 씌웁니다.

---
## 📝 과제 3 · 조건을 맞추면 회원은 몇 배인가

교재 4-2 3절에서 조건을 하나씩 더해 만든 `m3`와 같은 조건으로 회원의 배수를 구합니다. 조건은 `rain`, `C(hr)`, `temp_c`, `windspeed`, `C(mnth)`, `workingday`, `yr`입니다.

결과만 바꿉니다. 0건 시간을 빼지 않으려고 모든 시간의 회원 대여에 1을 더한 뒤 로그를 씌웁니다. 식의 왼쪽에 `np.log(registered + 1)`처럼 씁니다. 대여가 수십 건 이상인 시간에는 1을 더해도 로그 값이 거의 달라지지 않습니다.

- 적합한 모형은 `reg_model`, rain 계수에 `np.exp`를 씌운 배수는 `reg_ratio`에 담습니다.

**결과 형식:** `reg_model`은 적합한 모형, `reg_ratio`는 실수

In [11]:
# 📝 과제 3 — None 자리를 채우세요
# 힌트: 교재 4-2 3절 m3 셀의 두 줄에서 식의 왼쪽(결과)과 모형 이름만 바꿉니다.
reg_model = smf.ols(
    "np.log(registered + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + workingday + yr",
    data=h
).fit()  # 회원 대여(+1)의 로그에 교재 4-2 3절의 m3와 같은 조건을 넣은 모형
reg_ratio = np.exp(reg_model.params["rain"])  # rain 계수의 exp

In [12]:
# ✅ 확인 — 과제 3 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(3, reg_ratio=(reg_ratio, 0.001))
else:
    check_format(3, reg_model=reg_model, reg_ratio=reg_ratio)

형식 확인 완료: 과제 3 — 계산값의 정답 여부는 제출 후 채점합니다.


---
## 📝 과제 4 · 조건을 맞추면 비회원은 몇 배인가

과제 3과 같은 방법입니다. 결과만 `np.log(casual + 1)`로 바꿉니다. 두 집단을 같은 조건으로 맞춰야 배수를 비교할 수 있습니다.

- 모형은 `cas_model`, 배수는 `cas_ratio`에 담습니다.

**결과 형식:** `cas_model`은 적합한 모형, `cas_ratio`는 실수

In [13]:
# 📝 과제 4 — None 자리를 채우세요
# 힌트: 과제 3의 두 줄에서 registered를 casual로, 이름을 cas_로 바꿉니다.
cas_model = smf.ols(
    "np.log(casual + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + workingday + yr",
    data=h
).fit()  # 비회원 대여(+1)의 로그에 과제 3과 같은 조건을 넣은 모형
cas_ratio = np.exp(cas_model.params["rain"])  # rain 계수의 exp

In [14]:
# ✅ 확인 — 과제 4 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(4, cas_ratio=(cas_ratio, 0.001))
else:
    check_format(4, cas_model=cas_model, cas_ratio=cas_ratio)

형식 확인 완료: 과제 4 — 계산값의 정답 여부는 제출 후 채점합니다.


💬 **결과 읽기** — 과제 1의 평균끼리 배수와 비교해 보세요. 조건을 맞추자 회원과 비회원의 배수는 각각 어느 쪽으로 움직였나요? 두 집단의 차이는 커졌나요, 작아졌나요?

---
## ▶ 구간 확인 · 두 배수는 얼마나 흔들리나

교재 4-2 5절처럼 같은 날의 시간들을 한 묶음으로 보고, 과제 3·4에서 쓴 식 그대로 95% 신뢰구간을 계산해 배수로 바꿉니다. 계산 코드는 실행만 합니다.

In [15]:
# ▶ 그대로 실행하세요 — 과제 3·4 배수의 날짜로 묶은 95% 신뢰구간 (교재 4-2 5절과 같은 계산)
if reg_model is not None and cas_model is not None:
    days = h["dteday"].factorize()[0]
    intervals = {}
    for name, model in [("registered", reg_model), ("casual", cas_model)]:
        refit = smf.ols(model.model.formula, data=h).fit(cov_type="cluster", cov_kwds={"groups": days})
        intervals[name] = np.exp(refit.conf_int().loc["rain"]).values
    display(pd.DataFrame(intervals, index=["하한", "상한"]).round(3))
else:
    print("과제 3·4를 먼저 작성하세요.")

,registered,casual
하한,0.520,0.427
상한,0.605,0.518


💬 **결과 읽기** — 회원과 비회원의 구간을 나란히 보세요. 두 구간이 겹치는지만으로 두 배수의 차이를 검정했다고 말하지는 않습니다. 과제 7에서 차이를 얼마나 믿을 수 있는지 판단할 때 씁니다.

---
## 📝 과제 5 · 회원의 배수는 근무일과 주말·휴일에 다른가

회원과 비회원은 요일에 따라 대여하는 목적이 다를 수 있습니다. 교재 4-2 6절처럼 조건 필터로 근무일(`workingday` 1)과 주말·휴일(`workingday` 0)의 기록을 나누고, 과제 3의 식에서 `workingday`만 뺀 식을 각각 적합합니다. 한쪽 기록 안에서는 `workingday`가 모두 같기 때문입니다.

- 근무일의 배수는 `reg_ratio_wd`, 주말·휴일의 배수는 `reg_ratio_we`에 담습니다.

**결과 형식:** 실수 두 개

In [16]:
# 📝 과제 5 — None 자리를 채우세요
# 힌트: 교재 4-2 6절의 두 셀(기록 나누기 → 두 번 적합)과 같은 순서입니다. 식의 왼쪽은 과제 3과 같습니다.
h_wd = h[h["workingday"] == 1]
h_we = h[h["workingday"] == 0]

reg_model_wd = smf.ols(
    "np.log(registered + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + yr",
    data=h_wd
).fit()
reg_model_we = smf.ols(
    "np.log(registered + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + yr",
    data=h_we
).fit()

reg_ratio_wd =  np.exp(reg_model_wd.params["rain"])  # 근무일(workingday 1)만 적합한 회원 배수
reg_ratio_we = np.exp(reg_model_we.params["rain"])  # 주말·휴일(workingday 0)만 적합한 회원 배수

In [17]:
# ✅ 확인 — 과제 5 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(5, reg_ratio_wd=(reg_ratio_wd, 0.001), reg_ratio_we=(reg_ratio_we, 0.001))
else:
    check_format(5, reg_ratio_wd=reg_ratio_wd, reg_ratio_we=reg_ratio_we)

형식 확인 완료: 과제 5 — 계산값의 정답 여부는 제출 후 채점합니다.


---
## 📝 과제 6 · 비회원의 배수는 근무일과 주말·휴일에 다른가

과제 5와 같은 방법입니다. 결과만 `np.log(casual + 1)`로 바꿉니다.

- 근무일의 배수는 `cas_ratio_wd`, 주말·휴일의 배수는 `cas_ratio_we`에 담습니다.

**결과 형식:** 실수 두 개

In [18]:
# 📝 과제 6 — None 자리를 채우세요
# 힌트: 과제 5에서 나눈 두 기록을 그대로 쓰고, 식의 왼쪽과 변수 이름만 바꿉니다.

cas_model_wd = smf.ols(
    "np.log(casual + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + yr",
    data=h_wd
).fit()
cas_model_we = smf.ols(
    "np.log(casual + 1) ~ rain + C(hr) + temp_c + windspeed + C(mnth) + yr",
    data=h_we
).fit()

cas_ratio_wd = np.exp(cas_model_wd.params["rain"])  # 근무일(workingday 1)만 적합한 비회원 배수
cas_ratio_we = np.exp(cas_model_we.params["rain"])  # 주말·휴일(workingday 0)만 적합한 비회원 배수

In [19]:
# ✅ 확인 — 과제 6 (고치지 말고 실행만 하세요)
if os.environ.get("BALAB_ANSWER_KEY"):
    check(6, cas_ratio_wd=(cas_ratio_wd, 0.001), cas_ratio_we=(cas_ratio_we, 0.001))
else:
    check_format(6, cas_ratio_wd=cas_ratio_wd, cas_ratio_we=cas_ratio_we)

형식 확인 완료: 과제 6 — 계산값의 정답 여부는 제출 후 채점합니다.


과제 1~6의 배수를 한 표로 모읍니다. 과제를 모두 채워야 표가 나옵니다.

In [20]:
# ▶ 그대로 실행하세요 — 과제 1~6의 배수를 한 표로
if None not in (reg_mean_ratio, cas_mean_ratio, reg_ratio, cas_ratio,
                reg_ratio_wd, reg_ratio_we, cas_ratio_wd, cas_ratio_we):
    display(pd.DataFrame({
        "평균끼리 (과제 1)": [reg_mean_ratio, cas_mean_ratio],
        "조건을 맞춘 배수 (과제 3·4)": [reg_ratio, cas_ratio],
        "근무일 (과제 5·6)": [reg_ratio_wd, cas_ratio_wd],
        "주말·휴일 (과제 5·6)": [reg_ratio_we, cas_ratio_we],
    }, index=["registered", "casual"]).round(3))

,평균끼리 (과제 1),조건을 맞춘 배수 (과제 3·4),근무일 (과제 5·6),주말·휴일 (과제 5·6)
registered,0.601,0.561,0.551,0.584
casual,0.428,0.470,0.486,0.445


💬 **결과 읽기** — 행은 회원과 비회원, 열은 평균끼리·조건을 맞춘 배수·근무일·주말·휴일입니다. 가로로 읽으면 한 집단의 배수가 방법과 요일에 따라 어떻게 달라지는지, 세로로 읽으면 같은 기준에서 두 집단이 얼마나 다른지 보입니다.

---
## 📝 과제 7 · 본인의 판단과 설명

계산 결과를 근거로 두 집단의 계획 배수에 대한 본인의 판단을 씁니다. 사용한 수치를 함께 적습니다.

1. 회원과 비회원 중 비·눈에 더 크게 줄어드는 쪽은 어디인가요? 평균끼리의 배수와 조건을 맞춘 배수에서 판단이 같은가요? 다르다면 왜 다를 수 있을까요?
2. 날짜로 묶은 구간을 보고 두 집단의 배수 차이를 어느 정도 믿을 수 있는지 씁니다. 두 구간이 겹치는지만으로 차이를 검정했다고 말하지 않습니다.
3. 근무일과 주말·휴일에 따라 판단이 달라지나요? 이를 근거로 교재의 전체 배수 0.53 하나로 계획해도 되는지, 비회원이 많은 시간에는 달리 잡아야 하는지 본인의 입장을 씁니다.

자동 채점은 하지 않습니다.

✏️ **나의 답:단순히 평균끼리 비교했을 때 비·눈이 오는 경우 registered는 평시의 0.601배, casual은 0.428배로 나타났다. 다른 조건을 같게 맞추어 회귀분석한 결과에서는 registered는 0.561배, casual은 0.470배로 나타났다.  
근무일과 주말·휴일을 나누어 분석하면 registered는 근무일에 0.551배, 주말·휴일에 0.584배로, 비·눈의 영향이 근무일에 조금 더 크게 나타났다. 반면 casual은 근무일에 0.486배, 주말·휴일에 0.445배로, 주말·휴일에 비·눈의 영향이 더 크게 나타났다.  
따라서 단순 평균 비교와 다른 조건을 통제한 회귀분석의 결과에는 차이가 있으며, 근무일 여부에 따라서도 비·눈이 자전거 이용량에 미치는 영향이 다르게 나타난다는 것을 알 수 있다.** (여기를 두 번 눌러 적습니다)

---
## ✅ 제출 전 확인

아래 셀을 실행해 과제 1~6의 작성 여부와 형식을 확인합니다. "형식 확인 완료"는 계산값이 맞다는 뜻이 아닙니다. 결과를 직접 읽고 과제 7을 작성해 제출합니다. 실제 값 채점은 제출 후 진행합니다.

In [21]:
# ✅ 제출 전 확인 — 과제 1~6의 작성 여부와 형식
summary()

계산 과제 1~6 형식 확인 완료. 정답 판정은 아닙니다. 서술 답안도 작성해 제출하세요.


True